auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -0,0 +1,280 @@
+         */
+        private void bufferChangeEvent(JsonObject key, JsonObject changeEvent) {
+            LOG.info("Buffering change event for key {}", key);
+
+            BufferOffsets sequence = bufferOffsets().orElseGet(BufferOffsets::initial);
+
+            JsonObject wrapper = Json.createObjectBuilder()
+                    .add("key", key)
+                    .add("changeEvent", changeEvent)
+                    .build();
+            streamBuffer.put(sequence.getNextValueAndIncrement(), wrapper);
+
+            streamBuffer.put(BUFFER_OFFSETS_KEY, sequence.toJson());
+        }
+
+        /**
+         * Enriches the given change event with the metadata from the associated
+         * transaction.
+         *
+         * @return The enriched change event or {@code null} if no metadata for the
+         *         associated transaction was found.
+         */
+        private KeyValue<JsonObject, JsonObject> enrichWithTxMetaData(JsonObject key, JsonObject changeEvent) {
+            JsonObject txId = Json.createObjectBuilder()
+                    .add("transaction_id", changeEvent.get("source").asJsonObject().getJsonNumber("txId").longValue())
+                    .build();
+
+            JsonObject metaData = txMetaDataStore.get(txId);
+
+            if (metaData != null) {
+                LOG.info("Enriched change event for key {}", key);
+
+                metaData = Json.createObjectBuilder(metaData.get("after").asJsonObject())
+                        .remove("transaction_id")
+                        .build();
+
+                return KeyValue.pair(
+                        key,
+                        Json.createObjectBuilder(changeEvent)
+                            .add("audit", metaData)
+                            .build()
+                );
+            }
+
+            LOG.warn("No metadata found for transaction {}", txId);
+            return null;
+        }
+
+        private Optional<BufferOffsets> bufferOffsets() {
+            JsonObject bufferOffsets = streamBuffer.get(BUFFER_OFFSETS_KEY);
+            if (bufferOffsets == null) {
+                return Optional.empty();
+            }
+            else {
+                return Optional.of(BufferOffsets.fromJson(bufferOffsets));
+            }
+        }
+
+        @Override
+        public void close() {
+        }
+    }
+
+    /**
+     * Keeps track of the position within the key/value state store that is used as
+     * a buffer for non-joinable events. As there's no true queue like structure
+     * offered by Kafka Streams, a statestore is used, whose keys indicate the order
+     * of insertion and whose value are the buffered change events. This structure
+     * keeps track of the oldest insert position and the next insert position. It is
+     * stored in the same state store using a special key outside of the sequence
+     * range.
+     */
+    private static class BufferOffsets {
```

Comment:
@gunnarmorling This is very clever solution but I believe it complicates the example.
Looking at the code I can firmly claim that the store's iterator is ordered, the only problem is that the ordering is lexicographic by the key.
So to make it a little bit more idiomatic I propose to
 - introduce a logical timestamp or use a record metadata timestamp extractor to associate order id with each event
- use the keystore where the key will be the timestamp from above converted into string and left padded
- use get/set/all/range for keystore as needed as the order is guranteed

Reply:
@jpechane, that'd mean to rely on specifics of the `KeyValueStore` implementation, which isn't adviseable. The [docs](https://kafka.apache.org/23/javadoc/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.html#all--) of `all()` are very clear about this:

> No ordering guarantees are provided.

So things might work now, but may fail when using a different store implementation or when its implementation gets changed in a future version. Now one could do a *custom* statestore which provides the required ordering guarantees, but I think that'd even more complex than what's suggested here.

Reply:
@gunnarmorling I am sorry for being pesky but as I read the code you've written a queue implementation on top of keystore. Would not it be possible to use Kafka API and make a buffer in a topic?
So the semantics will be the same
 - new message arrives
 - is buffer empty?
   - yes
     - are tx metadata available?
       - yes - send the message
       - no - write message to the topic
   - no - read messages from buffer topic till associated metadata are available and send them downstream

Reply:
@jpechane, yes, that probably would work, too. Is it really worth the effort, though? That topic would have to be scoped to the specific KStreams instance, so it'd require some more thinking about the right partitioning strategy. The state store handles all this, plus it keeps the values in memory typically, so I'd assume it's better perf-wise, too. Keep in mind this buffer is usually empty or has a very small amount of entries.

So given I've already spent *way* more time on this than I had planned (and still need to write the blog post), I'd prefer to leave this as is for now. I can highlight the topic approach as an alternative implementation for the buffering in the blog post.

Reply:
@gunnarmorling Fine for me. I am just really surprised the is no OOTB implementation equivalent of `LinkedHashMap` as one would expect that processing in order is not an uncommon requirment. There is even `TimeOrderedKeyValueBuffer` but it is an internal only class.

Reply:
@jpechane @gunnarmorling my 2 cents on this: I pretty much agree with Gunnar here, especially when having the blog post in mind for which this example primarily serves at the moment. However, in general I wouldn't recommend to make any assumptions on the ordering of keys for the underlying / backing state store. After all, the whole idea of the state store abstraction is to allow *any* implementation. E.g. InMemoryKeyValueStore uses a TreeMap internally without a comparator option which results in the natural ordering of keys. Since keys are represented with the Bytes class this means there will be lexicographic key ordering. Relying on all of this facts is IMHO highest-possible coupling of a solution towards the underlying concrete implementation which is why I personally would want to avoid it at all cost.

Reply:
@hpgrahsl Thanks! Just what do you think about my idea to use topic instead of keyvalue store? I am not advocating to change the PR, just I'd like to validate the viability of the approach.

Reply:
I think it is a valid approach, and I'll mention it as one possible alternative in the blog post. One potential "gotcha" may be that any unmanaged Kafka producer employed by the streaming pipeline code wouldn't be part of the Kafka Streams "exactly once" semantics, which may or may not cause trouble.